airbyte-cdk/python/unit_tests/sources/declarative/test_factory.py

```diff
@@ -579,3 +633,177 @@ def test_add_fields(self):
+    content = """
+    stream_slicer:
+      type: DatetimeStreamSlicer
+      $options:
+        datetime_format: "%Y-%m-%dT%H:%M:%S.%f%z"
+      start_datetime:
+        type: MinMaxDatetime
+        datetime: "{{ config['start_time'] }}"
+        min_datetime: "{{ config['start_time'] + day_delta(2) }}"
+      end_datetime: "{{ config['end_time'] }}"
+      cursor_field: "created"
+      lookback_window: "5d"
+      start_time_option:
+        inject_into: request_parameter
+        field_name: created[gte]
+    """
+
+    config = parser.parse(content)
+    with pytest.raises(ValidationError):
+        factory.create_component(config["stream_slicer"], input_config, False)
+
+
+def test_validation_wrong_interface_type():
+    content = """
+    paginator:
+      type: "LimitPaginator"
+      page_size: 10
+      url_base: "https://airbyte.io"
+      limit_option:
+        inject_into: request_parameter
+        field_name: page_size
+      page_token_option:
+        inject_into: path
+      pagination_strategy:
+        type: "MinMaxDatetime"
+        datetime: "{{ response._metadata.next }}"
+    """
+    config = parser.parse(content)
+    with pytest.raises(ValidationError):
+        factory.create_component(config["paginator"], input_config, False)
+
+
+def test_validation_create_composite_error_handler():
+    content = """
+        error_handler:
+          type: "CompositeErrorHandler"
+          error_handlers:
+            - response_filters:
+                - predicate: "{{ 'code' in response }}"
+                  action: RETRY
+            - response_filters:
+                - http_codes: [ 403 ]
+    """
+    config = parser.parse(content)
+    with pytest.raises(ValidationError):
+        factory.create_component(config["error_handler"], input_config, False)
+
+
+# Leaving this test here to document a limitation of the validator. Decoder has no meaningful fields to validate on so it accepts
+# the MinMaxDatetime despite being the wrong type
+def test_validation_wrong_object_type():
+    content = """
+      paginator:
+        type: "LimitPaginator"
+        page_size: 10
+        url_base: "https://airbyte.io"
+        limit_option:
+          inject_into: request_parameter
+          field_name: page_size
+        page_token_option:
+          inject_into: path
+        pagination_strategy:
+          type: "CursorPagination"
+          cursor_value: "{{ response._metadata.next }}"
+        decoder:
+          type: "MinMaxDatetime"
+          datetime: "{{ response._metadata.next }}"
+    """
+    config = parser.parse(content)
+    factory.create_component(config["paginator"], input_config, False)
+
+
+# This test should fail because the extractor doesn't match the Array of resolved classes. However, despite the schema being correct
+# validation passes. Leaving this here to document it and revisit at another time. This is another validator limitation.
+def test_validate_types_nested_in_list():
```

Comment:
I've actually tested this by explicitly replacing the Interface in our code with a union of relevant types, but interestingly the validator still couldn't pick up on it.